// mjolnir.js/dist/hammerjs/recognizers/double-click-drag.js
import { Recognizer } from "../recognizer/recognizer.js";
import { RecognizerState } from "../recognizer/recognizer-state.js";
import { TOUCH_ACTION_MANIPULATION } from "../touchaction/touchaction-Consts.js";
import { InputEvent } from "../input/input-consts.js";
import { getPointDistance } from "../input/get-distance.js";
const EVENT_NAMES = ['', 'start', 'move', 'end', 'cancel'];
/**
 * DoubleClickDrag
 * Recognized when a second tap/click turns into a vertical drag gesture.
 */
export class DoubleClickDragRecognizer extends Recognizer {
    constructor(options = {}) {
        super({
            enable: true,
            event: 'doubleclickdrag',
            pointers: 1,
            interval: 500,
            time: 350,
            threshold: 28,
            dragThreshold: 1,
            pixelsPerScale: 120,
            ...options
        });
        this._tapStart = null;
        this._lastTap = null;
        this._drag = null;
        this._emittedStart = false;
    }
    getTouchAction() {
        return [TOUCH_ACTION_MANIPULATION];
    }
    getEventNames() {
        return EVENT_NAMES.map((suffix) => this.options.event + suffix);
    }
    process(input) {
        const { options } = this;
        const validPointers = input.pointers.length === options.pointers;
        if (!validPointers) {
            this.reset();
            return RecognizerState.Failed;
        }
        if (input.eventType & InputEvent.Start) {
            return this._handleStart(input);
        }
        if (input.eventType & InputEvent.Move) {
            return this._handleMove(input);
        }
        if (input.eventType & InputEvent.Cancel) {
            return this._handleEnd(input, true);
        }
        if (input.eventType & InputEvent.End) {
            return this._handleEnd(input, false);
        }
        return RecognizerState.Failed;
    }
    reset() {
        this._tapStart = null;
        this._lastTap = null;
        this._drag = null;
        this._emittedStart = false;
    }
    emit(input) {
        if (!input) {
            return;
        }
        if (this.state === RecognizerState.Began) {
            if (!this._drag?.active || this._emittedStart) {
                return;
            }
            this._emittedStart = true;
            this.manager.emit(`${this.options.event}start`, input);
            this.manager.emit(this.options.event, input);
            return;
        }
        if (this.state === RecognizerState.Changed) {
            if (!this._emittedStart) {
                return;
            }
            this.manager.emit(`${this.options.event}move`, input);
            this.manager.emit(this.options.event, input);
            return;
        }
        if (this.state === RecognizerState.Ended) {
            if (!this._emittedStart) {
                return;
            }
            this.manager.emit(this.options.event, input);
            this.manager.emit(`${this.options.event}end`, input);
            this._emittedStart = false;
            return;
        }
        if (this.state === RecognizerState.Cancelled) {
            if (!this._emittedStart) {
                return;
            }
            this.manager.emit(this.options.event, input);
            this.manager.emit(`${this.options.event}cancel`, input);
            this._emittedStart = false;
        }
    }
    _handleStart(input) {
        const pointerId = this._getPointerId(input);
        if (this._lastTap && this._isTapMatch(input, this._lastTap)) {
            this._tapStart = null;
            this._lastTap = null;
            this._drag = {
                startCenter: input.center,
                pointerId,
                active: false
            };
            this._emittedStart = false;
            return RecognizerState.Began;
        }
        this._tapStart = {
            center: input.center,
            timeStamp: input.timeStamp,
            pointerId
        };
        this._lastTap = null;
        this._drag = null;
        this._emittedStart = false;
        return RecognizerState.Failed;
    }
    _handleMove(input) {
        if (!this._drag || !this._isSamePointer(input, this._drag.pointerId)) {
            return RecognizerState.Failed;
        }
        const dy = this._drag.startCenter.y - input.center.y;
        if (!this._drag.active && Math.abs(dy) < this.options.dragThreshold) {
            return RecognizerState.Began;
        }
        this._drag.active = true;
        input.scale = Math.pow(2, dy / this.options.pixelsPerScale);
        return this._emittedStart ? RecognizerState.Changed : RecognizerState.Began;
    }
    _handleEnd(input, cancelled) {
        if (this._drag && this._isSamePointer(input, this._drag.pointerId)) {
            const { active, startCenter } = this._drag;
            this._drag = null;
            this._tapStart = null;
            this._lastTap = null;
            if (!active) {
                this._emittedStart = false;
                return RecognizerState.Failed;
            }
            const dy = startCenter.y - input.center.y;
            input.scale = Math.pow(2, dy / this.options.pixelsPerScale);
            return cancelled ? RecognizerState.Cancelled : RecognizerState.Ended;
        }
        if (!this._tapStart || !this._isSamePointer(input, this._tapStart.pointerId)) {
            if (cancelled) {
                this.reset();
            }
            return RecognizerState.Failed;
        }
        if (this._isValidTap(input)) {
            this._lastTap = {
                center: input.center,
                timeStamp: input.timeStamp,
                pointerId: this._tapStart.pointerId
            };
        }
        else {
            this._lastTap = null;
        }
        this._tapStart = null;
        return RecognizerState.Failed;
    }
    _isTapMatch(input, tap) {
        return (input.timeStamp - tap.timeStamp <= this.options.interval &&
            getPointDistance(input.center, tap.center) <= this.options.threshold);
    }
    _isValidTap(input) {
        return input.deltaTime <= this.options.time && input.distance <= this.options.threshold;
    }
    _getPointerId(input) {
        return 'pointerId' in input.srcEvent ? input.srcEvent.pointerId : null;
    }
    _isSamePointer(input, pointerId) {
        return pointerId === null || this._getPointerId(input) === pointerId;
    }
}
